CRM/Civicase/Hook/BuildForm/TokenTree.php

```diff
@@ -0,0 +1,392 @@
+      $newTokenTree[self::CURRENT_USER_TOKEN_TEXT]['children'][1]['children']
+        = array_values($newTokenTree[self::CURRENT_USER_TOKEN_TEXT]['children'][1]['children']);
+    }
+    if (!empty($newTokenTree[self::CLIENT_TOKEN_TEXT]['children'][0]['children'])) {
+      $newTokenTree[self::CLIENT_TOKEN_TEXT]['children'][0]['children']
+        = array_values($newTokenTree[self::CLIENT_TOKEN_TEXT]['children'][0]['children']);
+    }
+    if (!empty($newTokenTree[self::CLIENT_TOKEN_TEXT]['children'][1]['children'])) {
+      $newTokenTree[self::CLIENT_TOKEN_TEXT]['children'][1]['children']
+        = array_values($newTokenTree[self::CLIENT_TOKEN_TEXT]['children'][1]['children']);
+    }
+    if (!empty($newTokenTree[self::CASE_TOKEN_TEXT]['children'][0]['children'])) {
+      $newTokenTree[self::CASE_TOKEN_TEXT]['children'][0]['children']
+        = array_values($newTokenTree[self::CASE_TOKEN_TEXT]['children'][0]['children']);
+    }
+    if (!empty($newTokenTree[self::CASE_TOKEN_TEXT]['children'][1]['children'])) {
+      $newTokenTree[self::CASE_TOKEN_TEXT]['children'][1]['children']
+        = array_values($newTokenTree[self::CASE_TOKEN_TEXT]['children'][1]['children']);
+    }
+  }
+
+  /**
+   * Fetch all contact and case related custom fields.
+   */
+  private function fetchAllRelevantCustomFields() {
+    $customFields = [];
+    try {
+      $customFields = civicrm_api3('CustomField', 'get', [
+        'custom_group_id.extends' => [
+          'IN' => ['Contact', 'Individual', 'Household', 'Organization', 'Case'],
+        ],
+        'options' => ['limit' => 0],
+        'sequential' => 1,
+        'return' => ['id', 'custom_group_id.title'],
+      ]);
+    }
+    catch (Throwable $ex) {
+    }
+    if (!empty($customFields) && $customFields['is_error'] === 0) {
+      foreach ($customFields['values'] as $field) {
+        $this->customFields[$field['id']] = $field['custom_group_id.title'];
+      }
+    }
+  }
+
+  /**
+   * Add case role tokens to the new token tree.
+   *
+   * @param array $caseRoleTokens
+   *   Array of case role tokens.
+   * @param array $newTokenTree
+   *   Restructured token tree.
+   */
+  private function addCaseRoleTokens(array $caseRoleTokens, array &$newTokenTree) {
+    $contactRoleCount = 0;
+    $contactRoleTokens = [];
+    foreach ($caseRoleTokens as $key => $token) {
+      if ($token['id'] === '{case_roles.client}') {
+        continue;
+      }
+      $roleName = explode('-', $token['text']);
+      $roleName = $roleName[0];
+      $tokenName = 'Contact Role ' . $contactRoleCount . ' "' . trim($roleName) . '"';
+      if (empty($contactRoleTokens[$tokenName])) {
+        $contactRoleCount++;
+        $tokenName = 'Contact Role ' . $contactRoleCount . ' "' . trim($roleName) . '"';
+        $contactRoleTokens[$tokenName] = [
+          'id' => $this->clean($tokenName) . uniqid(),
+          'text' => $tokenName,
+          'children' => [],
+        ];
```

Comment:
Can we extract this intialization to a function where it can be used by other function that adds tokens as I notice those functions too do something similar

e.g

```php
  private function initializeTokenType(&$tokenTypeArray, $tokenName) {
    $tokenTypeArray[$tokenName] = [
      'id' => $this->clean($tokenName) . uniqid(),
      'text' => $tokenName,
      'children' => [],
    ];
  }

  private function initializeTokenTypeCoreField(&$tokenTypeArray, $tokenName, $token) {
    $tokenTypeArray[$tokenName]['children'][0] = [
      'id' => 'CoreFields' . uniqid(),
      'text' => 'Core Fields',
      'children' => [$token],
    ];
  }

```

Can you check if this will work?